Object State Transition Improvements

irq0 opened the pull request:
Improve and restrict object states:
- Open is treated as an intermediary state and no readily accessible by getters in the types.h abstraction
- Delete and commit transition functions get variants that allow pre conditions (if_state)
- Unversioned commit no longer deletes open versions to get last writer wins sematics (see commit for details)
- Objects with only open versions can be delete markered

Adds a unit test suite around types.h and the object state machine. The relevant modification functions now return a transaction successful boolean and bubble up to the types.h layer. Currently only the unit test suite checks this value. A follow up will tie this in with the SAL layer.

Fixes the dirty read problem for unversioned objects and adds the relevant s3 tests to our allowlist.

Fixes: https://github.com/aquarist-labs/s3gw/issues/462

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

Commit:
rgw/sfs: Change versioned object getter from non delete to committed

Change get versioned object logic from non deleted to only committed.
Essentially makes uncommited data invisible to types.h getters.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: createDBBucketBasic: create with deleted false

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: Add object state machine unit tests

Add new unit test suite that tests the object / versioned object state
machine from the types.h Object and Bucket API.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: conditional store object methods; basic error handling

Add store_versioned_object "if_state" variants that check if a version
is in a required state as a precondition to an update. Return boolean
success value and propagate that up the stack to the types.h delete
and metadata commit methods.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: Adapt unit test to only committed objects visible

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: Allow adding delete marker if latest version is open

Allow objects to get a delete marker if the latest version is not
yet committed.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
ci: Activate atomic write s3-tests

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: Only delete committed other versions

On unversioned commit we set other versions deleted. Before this
change this included open versions. Open versions, however, may be in
flight writes. Setting other in flight writes deleted will fail their
commit and we would essentially have first writer wins semantics. We
wont the last writer to win, so we only delete committed versions.

State changes now:
The object that we commit: open -> committed.
All committed -> deleted.
All open remain open.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

0xavi0 commented on src/rgw/driver/sfs/sqlite/sqlite_versioned_objects.cc:
```diff
@@ -107,6 +109,29 @@ void SQLiteVersionedObjects::store_versioned_object(
+    // throw exception (will be caught later in the sfs logic)
+    transaction.rollback();
+    return false;
+  }
```
Is this `count` needed?
`DBVersionObject::id` is unique, so you could get the object using the `id` and check the state in a helper function that iterates the vector of expected states that returns true/false.
I find it confusing (makes me think `id` is not unique when you `count where id = whatever` knowing the id in advance)

irq0 replied:
Agree that this is not the best solution. Follow up commit "rgw/sfs: conditional store: use update_all instead of select+update" changes this to update_all

0xavi0 commented on src/rgw/driver/sfs/sqlite/sqlite_versioned_objects.cc:
```diff
@@ -124,12 +149,54 @@ void SQLiteVersionedObjects::store_versioned_object_delete_rest_transact(
+      // throw exception (will be caught later in the sfs logic)
+      transaction.rollback();
+      return false;
+    }
```
Same as above.

irq0 replied:
Addressed in rgw/sfs: conditional store: use update_all instead of select+update

0xavi0 commented on src/rgw/driver/sfs/types.cc:
```diff
@@ -439,8 +447,11 @@ void Bucket::_delete_object_version(
   version.object_state = ObjectState::DELETED;
-  db_versioned_objs.store_versioned_object(version);
+  const bool ret = db_versioned_objs.store_versioned_object_if_state(
+      version, {ObjectState::OPEN, ObjectState::COMMITTED}
```
If get only returns objects with state = `COMMITTED`... how is it possible to delete an OPEN object?

`Object::delete_obj` calls `refresh_meta` --> `bucket->get()` --> `try_fetch_from_database` --> `get_committed_versioned_object`

The result of `bucket::get` is `throw UnknownObjectException();` and nothing is deleted.

For versioned buckets I suspect this race condition could end up creating a delete_marker and when the upload (the one OPEN) is done... it will commit the version with higher commit time, so it would prevail on top of the delete marker (because we get the last version based on the commit time)

This is the code that would end up creating the delete marker: (see comments with ***)
```c++
int SFSObject::SFSDeleteOp::delete_obj(
    const DoutPrefixProvider* dpp, optional_yield y
) {
  lsfs_dout(dpp, 10) << "bucket: " << source->bucket->get_name()
                     << " bucket versioning: "
                     << source->bucket->versioning_enabled()
                     << ", object: " << source->get_name()
                     << ", instance: " << source->get_instance() << dendl;

  // do the quick and dirty thing for now
  ceph_assert(bucketref);
  if (!source->objref) {
    source->refresh_meta(); // *** HERE WE REFRESH META, CALLS Bucket::get and returns nothing (exception) 
  }

  auto version_id = source->get_instance();
  std::string delete_marker_version_id;
  if (source->objref) { // *** source->objref is nullptr (Bucket::get returned nothing because of the exception)
    bucketref->delete_object(
        source->objref, source->get_key(), source->bucket->versioning_enabled(),
        delete_marker_version_id
    );
 // *** THIS WILL BE CALLED IF THE USER WAS DELETING THE OBJECT WITHOUT SPECIFYING A VERSION
 // *** A DELETE MARKER WILL BE CREATED
  } else if (source->bucket->versioning_enabled() && source->get_instance().empty()) {
    // create delete marker
    // even the object does not exist AWS creates a delete marker for it
    // if versioning is enabled and a specific version was not specified
    version_id =
        bucketref->create_non_existing_object_delete_marker(source->get_key());
  }
```

I agree with returning only COMMITTED objects (not open), but I'm not so sure about deleting OPEN objects.

I think we should differentiate when the object is OPEN, because in a race condition like the one described for versioned buckets we could end up with a non desired state.
If `Bucket::get` returns the OPEN object we can then differentiate and avoid the creation of that delete marker. 

I agree for reads only COMMITTED objects should be returned.

The problem is that S3 expects us to create that delete marker when deleting an object that does not exists.
There is an `s3test` for it.

I'm not sure what should we do in a race condition between UPLOAD vs DELETE and if we create the delete marker before... maybe we should delete it when the object is finally committed. 
I think it makes no sense to have a visible object that can also has a delete marker....

This could be a good topic to discuss on a dev call on Wednesdays. :)

irq0 replied:
I think we should not do the delete marker decision on OPEN versions.

OPEN versions are not guaranteed to transition to COMMITTED or DELETED. They may never leave OPEN when a client aborts a PUT for instance. Or we crash. 

While we store versions and delete markers in the same table with the same attributes, I think we must treat the *_time separate:
No:  t_v1 < t_v2 < t_deletemarker1 < t_v3 < t_deletemarker2
Yes: t_v1 < t_v2 < t_v3 < t_deletemarker1 < t_deletemarker2

My read of the spec is that delete markers are always on top of versions, regardless when they were set.

Commit:
rgw/sfs: conditional store: use update_all instead of select+update

Replace select, check, update with update_all and check in the where
clause. The select-check-update variant transaction needs a read write
upgrade, which during concurrent update may lead to transaction
aborts.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

0xavi0 commented on src/rgw/driver/sfs/sqlite/sqlite_versioned_objects.cc:
```diff
@@ -107,6 +109,33 @@ void SQLiteVersionedObjects::store_versioned_object(
+bool SQLiteVersionedObjects::store_versioned_object_if_state(
+    const DBVersionedObject& object, std::vector<ObjectState> allowed_states
+) const {
+  auto storage = conn->get_storage();
+  auto transaction = storage.transaction_guard();
+  storage.update_all(
+      set(c(&DBVersionedObject::object_id) = object.object_id,
+          c(&DBVersionedObject::checksum) = object.checksum,
+          c(&DBVersionedObject::size) = object.size,
+          c(&DBVersionedObject::create_time) = object.create_time,
+          c(&DBVersionedObject::delete_time) = object.delete_time,
+          c(&DBVersionedObject::commit_time) = object.commit_time,
+          c(&DBVersionedObject::mtime) = object.mtime,
+          c(&DBVersionedObject::object_state) = object.object_state,
+          c(&DBVersionedObject::version_id) = object.version_id,
+          c(&DBVersionedObject::etag) = object.etag,
+          c(&DBVersionedObject::attrs) = object.attrs,
+          c(&DBVersionedObject::version_type) = object.version_type),
+      where(
+          is_equal(&DBVersionedObject::id, object.id) and
+          in(&DBVersionedObject::object_state, allowed_states)
+      )
+  );
+  transaction.commit();
```
shouldn't `storage.changes()` be called before the `transaction.commit()`?
The examples I found state that in order to user `changes()` a transaction should be created, but it queries changes before returning from the lambda function that implements the transaction. 

For example:
```c++
storage.transaction([&] () mutable {
    storage.remove_all<User>(where(c(&User::id) < 100));
    auto usersRemoved = storage.changes();
    cout << "usersRemoved = " << usersRemoved << endl;
    return true;
});
```

If we have concurrent calls to this function and `changes` is not "inside" the transaction... could we get wrong values?

irq0 replied:
It should not matter, since changes is a call on the sqlite connection returning the number of changes of the last expression.
```
            int changes() {
                auto con = this->get_connection();
                return sqlite3_changes(con.get());
```

The reason there is an explicit transaction is because of sqlite_orm:

"It will print a number of deleted users (rows). But if you call `changes` without a transaction and your database **is located in file** not in RAM the result will be **0** always cause `sqlite_orm` opens and closes connection every time you call a function without a transaction."

That said, I just pushed the change (+ rebase). 
Better be consistent with the sqlite_orm example. Thanks for pointing that out!

0xavi0 reviewed (commented):
Just a question about the `changes` call.

Regarding the delete marker stuff, I guess it would be good to fix it in this PR as the changes in it introduce the possible error state, but I'm open to fix it in a separate PR (happy also to work on it)

irq0 commented:
> Regarding the delete marker stuff, I guess it would be good to fix it in this PR as the changes in it introduce the possible error state, but I'm open to fix it in a separate PR (happy also to work on it)

I think a follow up PR would make sense. Changing the delete marker sort logic to always on top *feels* like it deserves its own :)

0xavi0 reviewed (approved):
lg2m.
I've open a [bug](https://github.com/aquarist-labs/s3gw/issues/622), so we don't forget about the delete marker thing.
